### PR TITLE
[Network] Rebrand PeerMetadataStorage and modify API

### PR DESCRIPTION
**Pull request by JoshLind**
Note: More than 1k lines of code in this PR are simply new unit tests.

### Description
This PR makes the following changes:
- Renames `PeerMetadataStorage` to `PeersAndMetadata` (as the component will contain all peers and metadata information).
- Renames `PeerInfo` to `PeerMetadata` (as this is where we'll be storing all application specific metadata).
- Updates the API offered by `PeersAndMetadata` to be more comprehensive and cleaner. (This should make it much easier for applications, e.g., state sync, consensus, mempool, to interact with the component).
- Adds a comprehensive unit test suite for `PeersAndMetadata` (to make sure there are no gaps in the implementation. This is something that is currently missing).

To make this PR easier to review, we offer two commits:
1. Update the API offered by `PeersAndMetadata` and update existing application call sites.
2. Update the tests and add the new test suite. This is the bulk of the PR (>1k LoC).

I'd recommend reviewing each commit independently. It makes it obvious that there isn't a lot going on here except for the API changes and new tests.

### Test Plan
A new unit test suite and the existing test infrastructure.

### CODE REVIEW

**Review by grao1991 — APPROVED** [review]

File: network/src/application/storage.rs [hunk 1]
```diff
@@ -2,166 +2,199 @@
...
     collections::{hash_map::Entry, HashMap},
-    fmt::Debug,
     sync::Arc,
 };
 
-// TODO: refactor and clean up this interface.
-
-/// Metadata storage for peers across all of networking.  Splits storage of information across
-/// networks to prevent different networks from affecting each other
+/// A simple container that tracks all peers and peer metadata for the node.
```
Comment (ibalajiarun) [inline network/src/application/storage.rs:20]: mention that its grouped by NetworkId?

File: network/src/application/storage.rs [hunk 2]
```diff
@@ -2,166 +2,199 @@
...
+/// This container is updated by both the networking code (e.g., for new
+/// peer connections and lost peer connections), as well as individual
+/// applications (e.g., peer monitoring service).
 #[derive(Debug)]
-pub struct PeerMetadataStorage {
-    storage: HashMap<NetworkId, RwLock<HashMap<PeerId, PeerInfo>>>,
+pub struct PeersAndMetadata {
```
Comment (ibalajiarun) [inline network/src/application/storage.rs:25]: nit pick: the name says nothing about the NetworkId. Should it be `NetworkPeersAndMetadata`?

**Review by ibalajiarun — COMMENTED** [review]
LGTM, Just a nit pick on the name.

**Review by ibalajiarun — APPROVED** [review]
Approving to unblock.

**Review by sitalkedia — APPROVED** [review]
LGTM

File: network/src/application/storage.rs (hunk 2 above)
Comment (JoshLind) [reply network/src/application/storage.rs:25]: Interesting... I think `PeersAndMetadata` is okay, but yeah, maybe `NetworkPeersAndMetadata` is a little more obvious. I'll send out a PR once I flush the stack 😄

File: network/src/application/storage.rs (hunk 1 above)
Comment (JoshLind) [reply network/src/application/storage.rs:20]: Ideally we shouldn't need to expose this outside interface, i.e., the fact that it's grouped by the network id internally should be an implementation detail. Gonna leave for now and revisit once the API has stabilized.

**Pull request merged**